llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -6751,6 +6806,24 @@ Value *SPIRVToLLVM::transSPIRVImageWriteFromInst(SPIRVInstruction *bi, BasicBloc
   addr[lgc::Builder::ImageAddressIdxCoordinate] = transValue(bii->getOpValue(opndIdx++), bb->getParent(), bb);
   Value *texel = transValue(bii->getOpValue(opndIdx++), bb->getParent(), bb);
 
+  // Create smaller texel vector, if num texel components > num channels. This
+  // ensures that no unused channels are being stored.
+  if (auto texelVecType = dyn_cast<FixedVectorType>(texel->getType())) {
+    unsigned numChannels = getImageNumChannels(imageInfo.desc);
+    if (numChannels > 0 && numChannels < texelVecType->getNumElements()) {
+      if (numChannels == 1) {
+        texel = getBuilder()->CreateExtractElement(texel, getBuilder()->getInt32(0));
+      } else {
+        SmallVector<int, 3> channels;
```

Comment:
Could perhaps use `std::iota` here? I'm not sure how much it would help.

Reply:
This is a good suggestion, but a call to std::iota would operate on all reserved elements, except for std::iota(channels.begin(), channels.begin() + numChannels, 0), which I don't like that much. :-/

Reply:
Another option is to write `ArrayRef<int>({0, 1, 2, 3}).slice(0, numChannels)` instead of `channels`, but your version looks good already.